libc/string/string_lenchr.c

```diff
@@ -96,61 +99,215 @@ TEST(string_len, ascii)
+
+		/* Checking if we reverse order strcspn and strspn will find elements that are correct */
+		TEST_ASSERT_EQUAL_INT(BUFF_SIZE - 1, strcspn(&asciiStr[1], &reversCharSet[i]));
+
+		if (i == 127) {
+			TEST_ASSERT_EQUAL_INT(BUFF_SIZE - 1, strspn(&asciiStr[1], reversCharSet));
+		}
+		else {
+			TEST_ASSERT_EQUAL_INT(0, strspn(&asciiStr[1], reversCharSet));
+		}
+	}
+
+	free((void *)asciiStr);
+}
+
+
+TEST(string_spn, not_ascii)
+{
+	unsigned char charSet[BUFF_SIZE] = { 0 };
 	int sz, i;
 
-	/* The length of string is not clearly restricted, so we test one of bigger value, which may be used  */
-	for (i = 0; i < PATH_MAX - 1; i++) {
-		bigstr[i] = 'A';
+	/* Checking  out of ASCII bytes */
+	for (i = sizeof(charSet); i < sizeof(charSet) * 2 - 1; i++) {
+		charSet[i - sizeof(charSet)] = i;
 	}
-	bigstr[i] = '\0';
+
+	sz = sizeof(charSet) - 1;
+	TEST_ASSERT_EQUAL_INT(sz, strcspn((const char *)charSet, ""));
+	TEST_ASSERT_EQUAL_INT(sz, strspn((const char *)charSet, (const char *)charSet));
+}
+
+
+TEST(string_spn, big)
+{
+	char bigstr[PATH_MAX] = { 0 },
+		 support[PATH_MAX] = { 0 };
+	int sz;
+
+	/*
+	 * The length of the string is not restricted,
+	 * so we test one of the bigger values, which may be used
+	 * with remembering the last element must be a null term
+	 * zero to collect the data as the string from
+	 * declared space, that's why we memset one place less than its size
+	 */
 
 	sz = sizeof(bigstr) - 1;
-	TEST_ASSERT_EQUAL_INT(sz, strlen(bigstr));
-	TEST_ASSERT_EQUAL_INT(sz - 1, strnlen(bigstr, sz - 1));
-	TEST_ASSERT_EQUAL_INT(sz, strnlen(bigstr, sz));
-	TEST_ASSERT_EQUAL_INT(sz, strnlen(bigstr, sz + 1));
+	memset(bigstr, 'a', sz);
+
+	TEST_ASSERT_EQUAL_INT(sz, strcspn(bigstr, ""));
+	TEST_ASSERT_EQUAL_INT(sz, strspn(bigstr, bigstr));
+
+	memcpy(support, bigstr, sz);
+	support[sz - 3] = 'b';
+	TEST_ASSERT_EQUAL_INT(sz, strspn(bigstr, support));
+	support[sz - 3] = 'a';
+	TEST_ASSERT_EQUAL_INT(sz, strspn(bigstr, support));
+
+	memset(support, 'b', sz);
+	support[sz - 3] = 'a';
+	TEST_ASSERT_EQUAL_INT(0, strcspn(bigstr, support));
+	support[sz - 3] = 'b';
+	TEST_ASSERT_EQUAL_INT(sz, strcspn(bigstr, support));
+}
+
+
+TEST(string_spn, empty_output)
+{
+	TEST_ASSERT_EQUAL_INT(0, strcspn("", "abc"));
+	TEST_ASSERT_EQUAL_INT(0, strspn("", "abc"));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn("", ""));
+	TEST_ASSERT_EQUAL_INT(0, strspn("", ""));
+}
+
+
+TEST(string_spn, mixed_order)
+{
+	int i;
+	char *reversStr,
+		*testStr;
+
+	testStr = testdata_createCharStr(BUFF_SIZE);
+	reversStr = malloc(BUFF_SIZE + 1);
+
+	for (i = 1; i < BUFF_SIZE; i++) {
+		reversStr[i - 1] = testStr[BUFF_SIZE - i];
+	}
+
+	/* Creating bounds for strings operations remembering we start counting from 0 */
+	testStr[BUFF_SIZE - 1] = 0;
+	reversStr[BUFF_SIZE - 1] = 0;
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn("abc", "cba"));
+	TEST_ASSERT_EQUAL_INT(3, strspn("cba", "abc"));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn("abc", "bac"));
+	TEST_ASSERT_EQUAL_INT(3, strspn("bac", "abc"));
```

Comment:
The goal in this case is to test mixed order of `s2` with the same `s1` and assert that a behavior is proper. Mixing order in `s1` checks it too, but it's less readable in my opinion.

```suggestion
	TEST_ASSERT_EQUAL_INT(0, strcspn("abc", "cba"));
	TEST_ASSERT_EQUAL_INT(3, strspn("abc", "cba"));

	TEST_ASSERT_EQUAL_INT(0, strcspn("abc", "bac"));
	TEST_ASSERT_EQUAL_INT(3, strspn("abc", "bac"));

	TEST_ASSERT_EQUAL_INT(0, strcspn("abc", "bca"));
	TEST_ASSERT_EQUAL_INT(3, strspn("abc", "bca"));
```